rust/src/delta.rs

```diff
@@ -1048,15 +1058,100 @@ impl<'a> DeltaTransaction<'a> {
 
+    /// Add an arbitrary "action" to the actions associated with this transaction
+    pub fn add_action(&mut self, action: action::Action) {
+        self.actions.push(action);
+    }
+
+    /// Add an arbitrary number of actions to the actions associated with this transaction
+    pub fn add_actions(&mut self, actions: Vec<action::Action>) {
+        for action in actions.into_iter() {
+            self.actions.push(action);
+        }
+    }
+
+    /// Create a new add action and write the given bytes to the storage backend as a fully formed
+    /// Parquet file
+    ///
+    /// add_file accepts two optional parameters:
+    ///
+    /// partitions: an ordered vec of WritablePartitionValues for the file to be added
+    /// actions: an ordered list of Actions to be inserted into the log file _ahead_ of the Add
+    ///     action for the file added. This should typically be used for txn type actions
+    pub async fn add_file(
+        &mut self,
+        bytes: &Vec<u8>,
+        partitions: Option<Vec<(String, String)>>,
+    ) -> Result<(), DeltaTransactionError> {
+        let mut partition_values = HashMap::new();
+        if let Some(partitions) = &partitions {
+            for (key, value) in partitions {
+                partition_values.insert(key.clone(), value.clone());
+            }
+        }
+
+        let path = self.generate_parquet_filename(partitions);
+        let storage_path = self
+            .delta_table
+            .storage
+            .join_path(&self.delta_table.table_path, &path);
+
+        debug!("Writing a parquet file to {}", &storage_path);
+        self.delta_table
+            .storage
+            .put_obj(&storage_path, &bytes)
+            .await
+            .map_err(|source| DeltaTransactionError::Storage { source })?;
+
+        // Determine the modification timestamp to include in the add action - milliseconds since epoch
+        // Err should be impossible in this case since `SystemTime::now()` is always greater than `UNIX_EPOCH`
+        let modification_time = SystemTime::now().duration_since(UNIX_EPOCH).unwrap();
+        let modification_time = modification_time.as_millis() as i64;
+
+        self.actions.push(Action::add(action::Add {
+            path,
+            size: bytes.len() as i64,
+            partitionValues: partition_values,
+            partitionValues_parsed: None,
+            modificationTime: modification_time,
+            dataChange: true,
+            stats: None,
+            stats_parsed: None,
+            tags: None,
+        }));
+
+        Ok(())
+    }
+
+    fn generate_parquet_filename(&self, partitions: Option<Vec<(String, String)>>) -> String {
+        let mut path_parts = vec![];
+        /*
+         * The specific file naming for parquet is not well documented including the preceding five
+         * zeros and the trailing c000 string
+         *
+         */
+        path_parts.push(format!("part-00000-{}-c000.snappy.parquet", Uuid::new_v4()));
```

Comment:
typical hadoop parquet has a UUID, that is shared between all files in the batch. e.g. this useful in debugging.

instead of hardcoded `00000`, i'd ask to do an incrementing counter.

and instead of hard-coding `snappy`, i'd advice to take it from the compression engine used in `ParquetBuffer .. WriterProperties`.

Reply:
@nfx to summarize, this is the scheme we should be using for the partition name right?

`part-{COUNTER}-{BATCH_UUID}-c000.ext`.

Do you know what is the `c000` segment for?

> and instead of hard-coding snappy, i'd advice to take it from the compression engine used in ParquetBuffer .. WriterProperties.

I also agree with this one :+1:

Reply:
@houqp c000 might be the bucket number, but I don’t remember from top of my head. Will look once open a laptop next week.

i was writing custom parquet writer 4 years ago and researched it a bit then. It’s referenced from classes referenced in InsertIntoHadoopFsRelation in Apache Spark, for sure.. or at least not too far deep from it.

Main part: file naming won’t mean much for reading and writing (unless bucketed). It’s just for troubleshooting when something goes terribly wrong and you have to manually recover from fs. Probably won’t happen with delta 😜

Reply:
The counter stuff I'll file a ticket for later